syntax="proto3";

package bean.net;

option cc_generic_services = true;
option java_generic_services = true;
option py_generic_services = true;

enum MessageType 
{
    REQUEST = 0;
    RESPONSE = 1;
    ERROR = 2;
}

enum ErrorCode {
    NO_ERROR = 0;
    INVALID_REQUEST = 1;
    NO_SERVICE = 2;
    NO_METHOD = 3;
    INVALID_CHECKSUM = 4;
}

message RpcMessage
{
    MessageType type = 1;
    fixed64 id = 2;
    string service = 3;
    string method = 4;

    bytes request = 5;
    bytes response = 6;
    ErrorCode error = 7;
}